pkg/controller/nodeipam/ipam/multi_subnetworks_handler.go

```diff
@@ -0,0 +1,131 @@
+package ipam
+
+import (
+	"context"
+	"fmt"
+	nodetopologyv1 "github.com/GoogleCloudPlatform/gke-networking-api/apis/nodetopology/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilnode "k8s.io/cloud-provider-gcp/pkg/util/node"
+	"k8s.io/klog/v2"
+	"strings"
+)
+
+const nodeTopologyCRName = "default"
+
+func (ca *cloudCIDRAllocator) updateNodeTopology(node *v1.Node) error {
+	ctx := context.Background()
+
+	hasSubnetLabel, nodeSubnet := getNodeSubnetLabel(node)
+	if !hasSubnetLabel {
+		klog.V(2).Infof("Cannot find the subnet label on the node: %v", node.Name)
+	}
+
+	defaultSubnet, subnetPrefix, err := getSubnetWithPrefixFromURL(ca.cloud.SubnetworkURL())
+	if err != nil {
+		klog.Errorf("Error parsing the default subnetworkURL, err: %v", err)
+		return err
+	}
+
+	nodeTopologyCR, err := ca.nodeTopologyClient.NetworkingV1().NodeTopologies().Get(ctx, nodeTopologyCRName, metav1.GetOptions{})
+	if err != nil {
+		klog.Errorf("Failed to get NodeTopology: %v, err: %v", nodeTopologyCRName, err)
+		return err
+	}
+
+	crSubnets := nodeTopologyCR.Status.Subnets
+	if crSubnets == nil {
+		// Add the default subnet to the node topology CR
+		klog.V(2).Infof("No subnets found in the cr, adding the default subnet")
+		updatedCR := nodeTopologyCR.DeepCopy()
+		updatedCR.Status.Subnets = append(updatedCR.Status.Subnets, nodetopologyv1.SubnetConfig{
+			Name:       defaultSubnet,
+			SubnetPath: subnetPrefix + defaultSubnet,
+		})
+		// We always expect zones field in the status.
+		if updatedCR.Status.Zones == nil {
+			updatedCR.Status.Zones = []string{}
+		}
+		_, updateErr := ca.nodeTopologyClient.NetworkingV1().NodeTopologies().UpdateStatus(ctx, updatedCR, metav1.UpdateOptions{})
+		if updateErr != nil {
+			klog.Errorf("Error updating the CR: %v, err: %v", nodeTopologyCRName, updateErr)
+			return updateErr
+		}
+
+		klog.V(2).Infof("Successfully added the default subnet %v to nodetopology CR", defaultSubnet)
+
+		return nil
+	}
+
+	if !hasSubnetLabel {
+		klog.V(2).Infof("Default subnetwork is already updated in the CR.")
+		return nil
+	}
+
+	// Check if subnet already exists in the CR
+	for _, subnet := range crSubnets {
+		if subnet.Name == nodeSubnet {
+			klog.V(2).Infof("The subnet %s already exists in the node topology CR", nodeSubnet)
+			return nil
+		}
+	}
+
+	// We have a new subnet that should be added to the CR
+	// We assume all the subnets are in the same project and region
+	updatedCR := nodeTopologyCR.DeepCopy()
```

Comment:
deepcopy again? why we do not reuse from L40 when we might already did deepcopy?

Reply:
We always return in the block after writing the CR. This is a local copy.

Reply:
I am not sure if I understand the assumption. 
This code is always adding the default subnet to the CR because we cannot really remove this subnet from the cluster and it's good to always have this in the CR.

We make sure if the reconciliation logic as well that we always retain the default subnetwork of the cluster in the CR. So once we have added the it, we will always have at least one subnet.

Reply:
> I am not sure if I understand the assumption.

I was saying, if the current `nodeTopologyCR.Status.Subnets` is `nil`. And now I have an event coming on a non-default-subnet node. Then this code will add default subnet to `nodeTopologyCR.Status.Subnets` and returns. And we assume this is less likely to happen, or never happens.

Reply:
Can we add a short comment about this? It is not obvious to me in the beginning.